README.md
# Warthog GUI Wallet

![wallet overview](./docs/img/wallet.png)

> To use this wallet (sending txs and synchronizing balance) you'll need to provide a node to connect to. Either running locally on your machine or a node thats publicly accessible through the internet.

## Using Executables

### Requirements:

#### Windows
- [WebView2](https://go.microsoft.com/fwlink/p/?LinkId=2124703)
- [.NET](https://dotnet.microsoft.com/en-us/download/dotnet-framework/net462)

#### Linux
*tested for Ubuntu 22.04*
- `sudo apt install python3-gi python3-gi-cairo gir1.2-gtk-3.0 gir1.2-webkit2-4.1`

### Run:
Download executable for your operating system [here](https://github.com/andrewcrypto777/wart-wallet/releases).

The wallet will create a database file next to the executable when you run it.
It's best to put them in a folder like this:
```
└── MyWalletFolder
    ├── wartwallet.exe
    └── wartwallet.db
```

## Running From Source

### Requirements:
- Python 3.9 with required packages installed (requirements.txt)
- pyinstaller (if you want to bundle an executable yourself)

#### Linux only:
*tested for Ubuntu 22.04*
- `sudo apt install python3-gi python3-gi-cairo gir1.2-gtk-3.0 gir1.2-webkit2-4.1`
- `sudo apt install python3-venv python3-wheel python3-dev`
- ```
  sudo apt install libgirepository1.0-dev build-essential \
  libbz2-dev libreadline-dev libssl-dev zlib1g-dev libsqlite3-dev wget \
  curl llvm libncurses5-dev libncursesw5-dev xz-utils tk-dev libcairo2-dev
  ```
- `pip install PyGObject`

### Run:
`python3 main.py`

### Bundle executable yourself with pyinstaller:
To minimize executable size install requirements and run pyinstaller in a virtual environment.

*on linux* `sudo apt install binutils` *beforehand*

`pyinstaller main.spec`


CHANGELOG.md
- fix 

src/db.py
import sqlite3


class DB:
    def __init__(self):
        self.con = sqlite3.connect("../wartwallet.db", check_same_thread=False)
        with self.con:
            self.con.execute("""CREATE TABLE IF NOT EXISTS wallets (
                            id INTEGER PRIMARY KEY AUTOINCREMENT,
                            address TEXT,
                            pk TEXT,
                            salt TEXT,
                            last_balance TEXT NOT NULL DEFAULT '0'
                        );""")
            self.con.execute("""CREATE TABLE IF NOT EXISTS data (
                            id INTEGER PRIMARY KEY AUTOINCREMENT,
                            key TEXT,
                            value TEXT
                        );""")
            r = self.con.execute("SELECT COUNT() FROM data;")
            c = r.fetchone()[0]
            if c == 0:
                self.con.execute("INSERT INTO data (key, value) VALUES ('peer', 'http://localhost:3000');")

    def get_wallets(self):
        with self.con:
            r = self.con.execute("SELECT * FROM wallets;")
            data = r.fetchall()
            return data

    def get_wallet_by_address(self, addr):
        with self.con:
            r = self.con.execute(f"SELECT * FROM wallets WHERE address='{addr}';")
            data = r.fetchone()
            return data

    def insert_wallet(self, addr, pk, salt):
        with self.con:
            self.con.execute(f"INSERT INTO wallets (address, pk, salt) VALUES ('{addr}', '{pk}', '{salt}');")

    def update_balance(self, addr, balance):
        with self.con:
            self.con.execute(f"UPDATE wallets SET last_balance='{balance}' WHERE address='{addr}';")

    def delete_wallet(self, addr):
        with self.con:
            self.con.execute(f"DELETE FROM wallets WHERE address='{addr}';")

    def update_peer(self, peer):
        with self.con:
            self.con.execute(f"UPDATE data SET value='{peer}' WHERE key='peer';")

    def get_peer(self):
        with self.con:
            r = self.con.execute(f"SELECT * FROM data WHERE key='peer';")
            data = r.fetchone()
            return data[2]


src/utils.py
import base64
import os
import requests
import json
from cryptography.fernet import Fernet, InvalidToken
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from ecdsa import SigningKey, SECP256k1
from ecdsa.util import randrange_from_seed__trytryagain
from Crypto.Hash import RIPEMD160, SHA256
from pycoin.ecdsa.secp256k1 import secp256k1_generator
from hashlib import sha256
from random import randint


# encrypt private key
def encrypt_pk(pk, pw):
    salt = os.urandom(16)
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=480000,
    )
    key = base64.urlsafe_b64encode(kdf.derive(pw.encode()))
    f = Fernet(key)
    pk_enc = f.encrypt(pk.encode())
    # return pk and salt
    return pk_enc.decode(), base64.b64encode(salt).decode('utf-8')


# decrypt private key
def decrypt_pk(pk_enc, pw, salt):
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=base64.b64decode(salt),
        iterations=480000,
    )
    key = base64.urlsafe_b64encode(kdf.derive(pw.encode()))
    f = Fernet(key)
    try:
        pkhex = f.decrypt(pk_enc.encode())
    except InvalidToken:
        # wrong password
        return None
    return pkhex.decode()


def wallet_from_seed(seed):
    # generate private key
    secexp = randrange_from_seed__trytryagain(seed, SECP256k1.order)
    pk = SigningKey.from_secret_exponent(secexp, curve=SECP256k1)

    return address_from_pk(pk), pk


def wallet_from_pkhex(pkhex):
    # generate private key
    pk = SigningKey.from_string(bytes.fromhex(pkhex),curve=SECP256k1)

    return address_from_pk(pk), pk


def address_from_pk(pk):
    pubkey = pk.get_verifying_key().to_string('compressed')

    # convert public key to raw addresss
    sha = SHA256.SHA256Hash(pubkey).digest()
    addr_raw = RIPEMD160.RIPEMD160Hash(sha).digest()
    addr_raw.hex()

    # generate address by appending checksum
    addr_hash = SHA256.SHA256Hash(addr_raw).digest()
    checksum = addr_hash[0:4]
    addr = addr_raw + checksum

    return addr.hex()


def send_tx(recipient, amount, fee, pk, peer):
    # get pinHash and pinHeight from warthog node
    head_raw = requests.get(peer + '/chain/head').content
    head = json.loads(head_raw)
    pinHash = head["data"]["pinHash"]
    pinHeight = head["data"]["pinHeight"]

    # send parameters
    nonceId = randint(0, 4294967295)  # 32 bit number, unique per pinHash and pinHeight
    toAddr = recipient
    amountE8 = amount  # 1 WART, this must be an integer, coin amount * 10E8

    # alternative: round fee from E8 amount
    rawFeeE8 = str(fee)  # this needs to be rounded
    result = requests.get(peer + '/tools/encode16bit/from_e8/' + rawFeeE8).content
    encode16bit_result = json.loads(result)
    feeE8 = encode16bit_result["data"]["roundedE8"]  # 9992

    # generate bytes to sign
    to_sign = \
        bytes.fromhex(pinHash) + \
        pinHeight.to_bytes(4, byteorder='big') + \
        nonceId.to_bytes(4, byteorder='big') + \
        b'\x00\x00\x00' + \
        feeE8.to_bytes(8, byteorder='big') + \
        bytes.fromhex(toAddr)[0:20] + \
        amountE8.to_bytes(8, byteorder='big')

    # create signature
    private_key = pk.privkey.secret_multiplier
    digest = sha256(to_sign).digest()

    # sign with recovery id
    (r, s, rec_id) = secp256k1_generator.sign_with_recid(private_key, int.from_bytes(digest, 'big'))

    # normalize to lower s
    if s > secp256k1_generator.order() / 2:  #
        s = secp256k1_generator.order() - s
        rec_id ^= 1  # https://github.com/bitcoin-core/secp256k1/blob/e72103932d5421f3ae501f4eba5452b1b454cb6e/src/ecdsa_impl.h#L295
    signature65 = r.to_bytes(32, byteorder='big') + s.to_bytes(32, byteorder='big') + rec_id.to_bytes(1,
                                                                                                      byteorder='big')

    # post transaction request to warthog node
    postdata = {
        "pinHeight": pinHeight,
        "nonceId": nonceId,
        "toAddr": toAddr,
        "amountE8": amountE8,
        "feeE8": feeE8,
        "signature65": signature65.hex()
    }
    rep = requests.post(peer + "/transaction/add", json=postdata)
    return rep.json()


# redirect to /auth/<token>
def set_token(window):
    r = window.evaluate_js('window.location.replace("http://localhost:50050/auth/" + window.pywebview.token.toString());')


# get balance for address
def get_balance(peer, addr):
    try:
        r = requests.get(f"{peer}/account/{addr}/balance")
        data = r.json()
        if data["code"] == 0:
            return data["data"]["balance"]
        else:
            return 0
    except requests.exceptions.RequestException:
        return None


src/main.py
import threading
import sys
import webview
import re
from engineio.async_drivers import gevent
from flask import Flask, render_template, request, flash, redirect, url_for, make_response
from flask_apscheduler import APScheduler
from flask_socketio import SocketIO, emit
from mnemonic import Mnemonic
from utils import *
from secrets import compare_digest
from db import DB
import datetime

privkey = None
address = None
window = None
tx_buffer = None
balance_buffer = 0
con_status = "connecting"

mnemo = Mnemonic("english")

db = DB()

PEER = db.get_peer()

app = Flask(__name__)
app.secret_key = os.urandom(16).hex()

scheduler = APScheduler()
scheduler.init_app(app)
scheduler.start()

socketio = SocketIO(app)


@app.route('/')
def landing():
    return render_template("base.html")


# set token cookie to deny requests from outside of application
@app.route('/auth/<token>')
def auth(token):
    if not compare_digest(token, webview.token):
        return ""
    else:
        r = make_response(redirect(url_for('login')))
        r.set_cookie('auth', token)
        return r


# wallet selection
@app.route("/login", methods=["GET"])
def login():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    wallets = db.get_wallets()
    elements = ""
    for wallet in wallets:
        elements += f"<li><a href='/unlock/{wallet[1]}'>{wallet[1]}</a></li>"

    return render_template("login.html", wallets=elements)


# unlock wallet with password
@app.route("/unlock/<wallet>", methods=["GET", "POST"])
def unlock(wallet):
    global address, privkey, balance_buffer

    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        walletdata = db.get_wallet_by_address(wallet)
        decrypted_pk = decrypt_pk(walletdata[2], request.form.get("pw"), walletdata[3])
        if decrypted_pk is not None:
            addr, pk = wallet_from_pkhex(decrypted_pk)
            address = addr
            privkey = pk
            balance_buffer = walletdata[4]
            scheduler.get_job(id="sync").modify(next_run_time=datetime.datetime.now())
            return redirect(url_for('wallet'))
        else:
            flash('Invalid Password!')
            return redirect(request.url)
    else:
        return render_template("unlock.html")


# add new wallet
@app.route("/add")
def add():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    return render_template("add.html")


# choose password for restored wallet
@app.route("/restore-mnemo", methods=["GET", "POST"])
def restore_mnemo_pw():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        password1 = request.form.get("pw1")
        password2 = request.form.get("pw2")

        if password1 != password2:
            flash("Passwords are not the same!")
            return redirect(request.url)
        elif len(password1) < 8:
            flash("Password too short! (min. length: 8)")
            return redirect(request.url)

        return redirect(url_for("restore_mnemo", pw=password1))
    else:
        return render_template("new.html")


# input mnemonic for restored wallet
@app.route("/restore-mnemo/<pw>", methods=["GET", "POST"])
def restore_mnemo(pw):
    global address, privkey

    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        words = request.form.get("mnemo")
        seed = mnemo.to_seed(words, passphrase="")
        address, privkey = wallet_from_seed(seed)

        encrypted_pk, salt = encrypt_pk(privkey.to_string().hex(), pw)

        db.insert_wallet(address, encrypted_pk, salt)
        scheduler.get_job(id="sync").modify(next_run_time=datetime.datetime.now())

        return redirect(url_for("wallet"))
    else:
        return render_template("restore-mnemo.html")


# choose password for restored wallet
@app.route("/restore-pk", methods=["GET", "POST"])
def restore_pk_pw():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        password1 = request.form.get("pw1")
        password2 = request.form.get("pw2")

        if password1 != password2:
            flash("Passwords are not the same!")
            return redirect(request.url)
        elif len(password1) < 8:
            flash("Password too short! (min. length: 8)")
            return redirect(request.url)

        return redirect(url_for("restore_pk", pw=password1))
    else:
        return render_template("new.html")


# input pk for restored wallet
@app.route("/restore-pk/<pw>", methods=["GET", "POST"])
def restore_pk(pw):
    global address, privkey

    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        pkhex = request.form.get("pk")
        address, privkey = wallet_from_pkhex(pkhex)

        encrypted_pk, salt = encrypt_pk(privkey.to_string().hex(), pw)

        db.insert_wallet(address, encrypted_pk, salt)
        scheduler.get_job(id="sync").modify(next_run_time=datetime.datetime.now())

        return redirect(url_for("wallet"))
    else:
        return render_template("restore-pk.html")


# generate new wallet
@app.route("/new", methods=["GET", "POST"])
def new():
    global address, privkey

    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        password1 = request.form.get("pw1")
        password2 = request.form.get("pw2")

        if password1 != password2:
            flash("Passwords are not the same!")
            return redirect(request.url)
        elif len(password1) < 8:
            flash("Password too short! (min. length: 8)")
            return redirect(request.url)

        words = mnemo.generate(strength=256)
        seed = mnemo.to_seed(words, passphrase="")
        address, privkey = wallet_from_seed(seed)

        encrypted_pk, salt = encrypt_pk(privkey.to_string().hex(), password1)

        db.insert_wallet(address, encrypted_pk, salt)
        scheduler.get_job(id="sync").modify(next_run_time=datetime.datetime.now())

        return redirect(url_for("new_confirm", mnemonic=words))

    else:

        return render_template("new.html")


# display mnemonic for new wallet
@app.route("/new-confirm/<mnemonic>")
def new_confirm(mnemonic):
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    return render_template("new-confirm.html", mnemonic=mnemonic)


# wallet dashboard
@app.route("/wallet")
def wallet():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    return render_template("wallet.html", address=address)


# send transaction
@app.route("/send", methods=["GET", "POST"])
def send():
    global tx_buffer

    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        recipient = request.form.get("recipient")
        amount = request.form.get("amount")
        fee = request.form.get("fee")

        if not re.search("^[a-fA-F0-9]{48}$", recipient):
            flash(f"Invalid recipient.")
            return render_template("send.html")

        try:
            amount = float(amount)
            fee = float(fee)
        except ValueError:
            flash(f"Invalid value.")
            return render_template("send.html")

        amount_int = round(amount * 10 ** 8)
        fee_int = round(fee * 10 ** 8)

        tx_buffer = [recipient, amount_int, fee_int, privkey]

        flash(f"You are sending {amount:.8f} WART to {recipient}.<div class='center'><a class='btn' id='send-confirm' href='/submit'>Confirm</a></div>")
        return render_template("send.html")
    else:
        return render_template("send.html")


# submit transaction to peer
@app.route("/submit")
def submit():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    emit("sending", namespace="/", broadcast=True)

    try:
        r = send_tx(tx_buffer[0], tx_buffer[1], tx_buffer[2], tx_buffer[3], PEER)
        if r['code'] == 0:
            flash('Success!')
            return redirect(url_for("wallet"))
        else:
            flash(str(r))
            return redirect(url_for("send"))
    except requests.exceptions.RequestException as e:
        flash(f"Connection error. Try changing your peer in the settings.")
        return redirect(url_for("send"))


# settings
@app.route("/settings", methods=["GET", "POST"])
def settings():
    global PEER

    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    if request.method == "POST":
        PEER = request.form.get("peer")
        db.update_peer(PEER)
        flash("Settings saved!")
        return render_template("settings.html", peer=PEER)
    else:
        return render_template("settings.html", peer=PEER)


# delete confirm
@app.route("/delete")
def delete_confirm():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    return render_template("delete.html", address=address)


# delete wallet
@app.route("/delete/<address>")
def delete_wallet(address):
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""

    db.delete_wallet(address)
    return redirect(url_for("login"))


# exit application
@app.route("/exit")
def exit_app():
    token = request.cookies.get('auth')
    if not compare_digest(str(token), webview.token):
        return ""
    if address is not None:
        db.update_balance(address, balance_buffer)
    window.destroy()
    sys.exit()


# sync
@scheduler.task('interval', id='sync', seconds=10, misfire_grace_time=20)
def sync():
    global con_status

    with scheduler.app.app_context():
        global balance_buffer
        if address is not None:
            b = get_balance(PEER, address)
            if b is not None:
                balance_buffer = b
                con_status = "connected"
            else:
                con_status = "disconnected"
            print("Synced")


# socketio
@socketio.on('connect')
def first_connect():
    balances = f"""<p class="wallet-amount">{balance_buffer} WART</p><p class="wallet-amountusd">0 USD</p>"""
    emit("overview", {"data": balances}, namespace="/")
    emit("con", {"data": con_status}, namespace="/")


@socketio.on('update_wallet')
def update_wallet():
    balances = f"""<p class="wallet-amount">{balance_buffer} WART</p><p class="wallet-amountusd">0 USD</p>"""
    emit("overview", {"data": balances}, namespace="/")
    emit("con", {"data": con_status}, namespace="/")


def start_server():
    socketio.run(app, host='localhost', port=50050)


if __name__ == '__main__':
    t = threading.Thread(target=start_server)
    t.daemon = True
    t.start()

    window = webview.create_window("WART Wallet", "http://localhost:50050/", width=1280, height=720, text_select=True,
                                   resizable=False, background_color="#212121", frameless=True, easy_drag=False)
    webview.start(set_token, window)
    sys.exit()
